src/backend/distributed/shardsplit/shardsplit_decoder.c

```diff
@@ -129,6 +134,40 @@ split_change_cb(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
 	}
 
 	Relation targetRelation = RelationIdGetRelation(targetRelationOid);
+
+	/*
+	 * If any columns from source relation have been dropped, then the tuple needs to
+	 * be formatted according to the target relation.
+	 */
+	TupleDesc sourceRelationDesc = RelationGetDescr(relation);
+	TupleDesc targetRelationDesc = RelationGetDescr(targetRelation);
+	if (sourceRelationDesc->natts > targetRelationDesc->natts)
+	{
+		HeapTuple targetRelationTuple = GetTupleForTargetSchema(existingTuple,
+																sourceRelationDesc,
+																targetRelationDesc);
+		switch (change->action)
+		{
+			case REORDER_BUFFER_CHANGE_INSERT:
+			{
+				change->data.tp.newtuple->tuple = *targetRelationTuple;
+				break;
+			}
+
+			case REORDER_BUFFER_CHANGE_UPDATE:
+			{
+				change->data.tp.newtuple->tuple = *targetRelationTuple;
```

Comment:
what about the old tuple? (might only have the replica identity columns)

Reply:
@marcocitus : In the UPDATE, we are only considering the new tuple which has the updated values. This updated tuple should actually be according to the schema of source shard(which means it would have some dropped columns too)

Reply:
The old tuple is also according to the schema of the old shard so it could have dropped columns. The old tuple is used to find which tuple to update on the target, which does not have the dropped column.

Reply:
Got it. Thanks for the explanation. So this means that we should also format the old tuple? In the existing testcase, I can try to see if UPDATE is working as expected without this change.

Reply:
> Got it. Thanks for the explanation. So this means that we should also format the old tuple? In the existing testcase, I can try to see if UPDATE is working as expected without this change.

yes